### ISSUE TITLE
Execute actions programatically

### ISSUE THREAD

**Issue by kurnevsky** [oantolin/embark#477]
It would be nice to have a way to execute some action without triggering embark prompt. For instance, I'd like to bind `C-k` to `kill-buffer` while executing `consult-buffer`. Is it possible to do so?

**Comment by minad** [comment]
You can bind a keyboard macro to the key C-k.

**Comment by kurnevsky** [comment]
Well, I'd consider this a hack :)
But I can live with this if there is no other way and it is not going to be implemented.
Since @minad you are here, is it possible to kill a buffer from `consult-buffer` and don't change buffer ordering? When I kill it with embark without closing minibuffer the ordering is changed - buffer above the killed one moves to the first position together with the pointe. I'm trying to replace `ivy-switch-buffer` with `consult-buffer`, and `ivy-switch-buffer` has really convenient way of killing :)

**Comment by oantolin** [comment]
You don't need Embark for this at all:

```elisp
(defun kill-target-buffer ()
  (interactive)
  (kill-buffer (vertico--candidate)))
```

Substitute the `(vertico--candidate)` for the appropriate expression for whatever completion UI you use.

You can, if you wish, reuse Embark's target picking mechanism:

```elisp
(defun kill-target-buffer ()
  (interactive)
  (if-let ((buffer (seq-some
                      (lambda (target)
                        (when (eq (plist-get target :type) 'buffer)
                          (plist-get target :target)))
                      (embark--targets))))
      (kill-buffer buffer)
    (user-error "No buffer target found")))
```

That way you can kill buffer in ibuffer too, for example.

**Comment by kurnevsky** [comment]
> You don't need Embark for this at all:

This function has to be a bit more complex:

```el
(let* ((candidate (vertico--candidate))
    (name (substring candidate 0 (1- (length candidate)))))
  (kill-buffer name))
```

For some reason candidate contains some unicode character additionally to buffer name.

Also it doesn't work that smoothly - vertico doesn't delete buffer from the candidates list when it's killed.

**Comment by minad** [comment]
@kurnevsky Yes, these issues that you observe with the programmatic solution are all expected. Embark does a bit more than the minimal `kill-target-buffer` function by @oantolin. They are solved if you use Embark instead via a keyboard macro. The name Emacs is an acronym "Editor MACroS". Using keyboard macros is not a hack. Why do you believe that? It is the most direct solution to shorten key sequences, in this case `C-k = C-. k`.

**Comment by minad** [comment]
@oantolin Is there an easy way to write the equivalent to the `C-. k` keyboard macro as an Elisp function? I think this is not even that trivial due to the control flow of `embark-act`.

**Comment by oantolin** [comment]
Yeah, keyboard macro is easier, but this function is not too awful, I hope:

```elisp
(defun kill-target-buffer ()
  (interactive)
  (if-let ((buffer (seq-find
                      (lambda (target)
                        (eq (plist-get target :type) 'buffer))
                      (embark--targets))))
      (embark--act 'kill-buffer buffer)
    (user-error "No buffer target found")))
```

Using `embark--targets` ensures all target finders _and_ candidate transformers are applied, and using `embark--act` ensures all pre-action, target injection and post-action hooks are run (particularly `embark--restart` in this case).

**Comment by kurnevsky** [comment]
Thanks everyone! I like this `kill-target-buffer` function :)
The only problem left is that buffer preview messes up buffer ordering when `embark--restart` is executed. But previews turned off it works almost perfectly :)

**Comment by oantolin** [comment]
I had completely forgotten, but I had written about this topic [on reddit about 7 months ago](https://www.reddit.com/r/emacs/comments/pac8kp/how_to_send_an_embark_command_without_pulling_up/).

**Comment by hmelman** [comment]
> I had completely forgotten, but I had written about this topic [on reddit about 7 months ago](https://www.reddit.com/r/emacs/comments/pac8kp/how_to_send_an_embark_command_without_pulling_up/).

That's a really good comment.  I suggest putting it in the embark README or an FAQ or at least the wiki.
